src/sinks/datadog/logs/sink.rs

```diff
@@ -0,0 +1,376 @@
+        let (members, finalizers) = self.dissect_batch(batch);
+
+        let total_members = members.len();
+        assert!(total_members <= MAX_PAYLOAD_ARRAY);
+        let body: Vec<u8> = serde_json::to_vec(&members).expect("failed to encode to json");
+        let serialized_payload_bytes_len = body.len();
+        metrics::histogram!(
+            "encoded_payload_size_bytes",
+            serialized_payload_bytes_len as f64
+        );
+        let (encoded_body, is_compressed) = match self.compression {
+            Compression::None => (body, false),
+            Compression::Gzip(level) => {
+                let level = level.unwrap_or(GZIP_FAST);
+                let mut encoder = GzEncoder::new(
+                    Vec::with_capacity(serialized_payload_bytes_len),
+                    flate2::Compression::new(level as u32),
+                );
+
+                encoder.write_all(&body)?;
+                (encoder.finish().expect("failed to encode"), true)
+            }
+        };
+        Ok(LogApiRequest {
+            serialized_payload_bytes_len,
+            payload_members_len: total_members,
+            api_key: Arc::clone(&api_key),
+            is_compressed,
+            body: encoded_body,
+            finalizers,
+        })
+    }
+}
+
+#[async_trait]
+impl<S> StreamSink for LogSink<S>
+where
+    S: Service<LogApiRequest> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Response: Response + Send + 'static,
+    S::Error: Debug + Into<crate::Error> + Send,
+{
+    async fn run(&mut self, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let io_bandwidth = 64;
+        let (io_tx, io_rx) = channel(io_bandwidth);
+        let service = self
+            .service
+            .take()
+            .expect("same sink should not be run twice");
+        let acker = self
+            .acker
+            .take()
+            .expect("same sink should not be run twice");
+        let encoding = self
+            .encoding
+            .take()
+            .expect("same sink should not be run twice");
+        let default_api_key = Arc::clone(&self.default_api_key);
+        let compression = self.compression;
+        let log_schema = self.log_schema;
+
+        let io = run_io(io_rx, service, acker).in_current_span();
+        let _ = tokio::spawn(io);
+
+        let batcher = Batcher::new(
+            input,
+            EventPartitioner::default(),
+            self.timeout,
+            NonZeroUsize::new(MAX_PAYLOAD_ARRAY).unwrap(),
+            NonZeroUsize::new(MAX_PAYLOAD_BYTES),
+        )
+        .map(|(maybe_key, batch)| {
+            let key = maybe_key.unwrap_or_else(|| Arc::clone(&default_api_key));
+            let request_builder = RequestBuilder::new(encoding.clone(), compression, log_schema);
+            tokio::spawn(async move { request_builder.build(key, batch) })
+        })
+        .buffer_unordered(io_bandwidth);
```

Comment:
Between this and the IO channel, will we end up with 128 built requests buffered before we propagate backpressure from the service itself? That's not necessarily bad, we'll just want to keep in mind memory use, responsiveness, etc.

Reply:
It's true. The user's in-memory limit is, so far as I'm aware, soft in our documentation so it's within our constraints to add this padding but I don't _love_ it. Similar thing in the s3 sink(s). Maybe we do a follow-up PR to deduct any internal buffering from the batch limit set by the user? Implies a need to set a minimal batch size at the config level I think. 

FWIW in practice I can't get this buffer to actual populate all that much. Every so often I wonder what a fixed-memory vector would look like and these would sure need to count against the global memory.